src/common/PopupContainerViewBase.tsx

```diff
@@ -28,6 +29,249 @@ export interface PopupComponent {
+                    break;
+
+                case 'right':
+                    absX = anchorRect.right;
+                    absY = anchorRect.top + (anchorRect.height - popupRect.height) / 2;
+                    anchorOffset = popupRect.height / 2;
+
+                    if (popupRect.width <= windowDims.width - ALLEY_WIDTH - anchorRect.right) {
+                        foundPerfectFit = true;
+                    } else if (!foundPartialFit) {
+                        constrainedWidth = windowDims.width - ALLEY_WIDTH - anchorRect.right;
+                    }
+                    break;
+
+                case 'left':
+                    absX = anchorRect.left - popupRect.width;
+                    absY = anchorRect.top + (anchorRect.height - popupRect.height) / 2;
+                    anchorOffset = popupRect.height / 2;
+
+                    if (popupRect.width <= anchorRect.left - ALLEY_WIDTH) {
+                        foundPerfectFit = true;
+                    } else if (!foundPartialFit) {
+                        constrainedWidth = anchorRect.left - ALLEY_WIDTH;
+                    }
+                    break;
+
+                case 'context':
+                    // Search for perfect fits on the LR, LL, TR, and TL corners.
+                    const fitsAbove = anchorRect.top - popupRect.height >= ALLEY_WIDTH;
+                    const fitsBelow = anchorRect.top + anchorRect.height + popupRect.height <= windowDims.height - ALLEY_WIDTH;
+                    const fitsLeft = anchorRect.left - popupRect.width >= ALLEY_WIDTH;
+                    const fitsRight = anchorRect.left + anchorRect.width + popupRect.width <= windowDims.width - ALLEY_WIDTH;
+                    if (fitsBelow && fitsRight) {
+                        foundPerfectFit = true;
+                        absX = anchorRect.left + anchorRect.width;
+                        absY = anchorRect.top + anchorRect.height;
+                    } else if (fitsBelow && fitsLeft) {
+                        foundPerfectFit = true;
+                        absX = anchorRect.left - popupRect.width;
+                        absY = anchorRect.top + anchorRect.height;
+                    } else if (fitsAbove && fitsRight) {
+                        foundPerfectFit = true;
+                        absX = anchorRect.left + anchorRect.width;
+                        absY = anchorRect.top - popupRect.height;
+                    } else if (fitsAbove && fitsLeft) {
+                        foundPerfectFit = true;
+                        absX = anchorRect.left - popupRect.width;
+                        absY = anchorRect.top - popupRect.height;
+                    }
+                    break;
+            }
+
+            const effectiveWidth = constrainedWidth || popupRect.width;
+            const effectiveHeight = constrainedHeight || popupRect.height;
+
+            // Make sure we're not hanging off the bounds of the window.
+            if (absX < ALLEY_WIDTH) {
```

Comment:
None of these adjustments are made in the 'context' case. Is this intended? What happens if there is no "perfect fit" and the context menu hangs off the edge of the window? Don't you want to nudge it back on and/or resize it so it fits?